kolibri/core/content/utils/annotation.py

```diff
@@ -30,109 +34,368 @@
+            )
+            # Add this statement to the query
+            batch_statement = batch_statement.where(
+                ContentNodeTable.c.id.in_(node_ids_statement)
+            )
+
+        if exclude_node_ids is not None:
+            # Construct a statement that restricts nodes we update
+            # in this batch by the specified exclusion constraints
+            exclude_node_ids_statement = _MPTT_descendant_ids_statement(
+                connection,
+                ContentNodeTable,
+                exclude_node_ids,
+                min_boundary,
+                max_boundary,
+            )
+            # Add this statement to the query
+            batch_statement = batch_statement.where(
+                ~ContentNodeTable.c.id.in_(exclude_node_ids_statement)
+            )
+
+        # Execute the update for this batch
+        connection.execute(
+            batch_statement.values(available=False).execution_options(autocommit=True)
         )
 
-    connection.execute(
-        update_statement.values(available=False).execution_options(autocommit=True)
-    )
+        i += 1
 
     bridge.end()
 
 
 def set_leaf_node_availability_from_local_file_availability(
     channel_id, node_ids=None, exclude_node_ids=None
 ):
+    """
+    Set nodes in a channel as available, based on their required files.
+    With no additional arguments, this will make every node in the channel
+    available or unavailable based on whether the files needed to render
+    those nodes are present on disk.
+    With the additional nodes arguments, it will selectively flag nodes
+    based on the passed in ids, marking their availability if
+    they are in node_ids, or descendants of those nodes, but not in
+    exclude_node_ids or descendants of those nodes.
+    Nodes in the channel not captured by the constraints will not have
+    their availability changed either way.
+    """
     bridge = Bridge(app_name=CONTENT_APP_NAME)
 
+    # SQL Alchemy reference to the content node table
     ContentNodeTable = bridge.get_table(ContentNode)
+    # SQL Alchemy reference to the file table - a mapping from
+    # contentnodes to the files that they use
     FileTable = bridge.get_table(File)
+    # SQL Alchemy reference to the localfile table which tracks
+    # information about the files on disk, such as availability
     LocalFileTable = bridge.get_table(LocalFile)
 
     connection = bridge.get_connection()
 
+    # This statement defines the update condition for the contentnode
+    # running exists on this (as it is used below) will produce either
+    # True, in the case when the contentnode has the required files
+    # available for rendering, or False otherwise.
     contentnode_statement = (
-        select([FileTable.c.contentnode_id])
-        .select_from(
+        # We could select any property here, as it's the exist that matters.
+        select([FileTable.c.contentnode_id]).select_from(
```

Comment:
Note that you can also just do things like `SELECT 1 FROM kolibriauth_facilityuser;` (maybe faster...?)

Reply:
Example here: https://stackoverflow.com/questions/7646173/sqlalchemy-exists-for-query#comment65330050_13336408